src/transformers/modeling_tf_pytorch_utils.py

```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
+
+
+def _load(f, shapes, prefix=""):
+    """
+    Recursive internal function to load all the names of the tensors included in an h5 file
+    """
+    import h5py
+
+    for k in f.keys():
+        if isinstance(f[k], h5py._hl.dataset.Dataset):
+            shapes[f"{prefix}/{k}"] = f[k].shape
+        else:
+            _load(f[k], shapes, prefix=f"{prefix}/{k}")
+
+
+def load_h5_weight_names(resolved_archive_file) -> Dict[str, Shape]:
+    """
+    Load all names and shapes of tensors included in a given TF file.
+    """
+    import h5py
+
+    # Read the H5 file
+    shapes = {}
+    with h5py.File(resolved_archive_file, "r") as f:
+        # Retrieve the name of each layer from the H5 file
+        _load(f, shapes)
+    return shapes
+
+
+def load_single(f, fullname):
+    """
+    Given an h5 file (as file handle), returns the data under the `fullname` location.
+    """
+    names = fullname.strip("/").split("/")
+
+    # Retrieve the name of each layer from the H5 file
+    data = f
+    for name in names:
+        data = data[name]
+    return data
+
+
+def set_weight(pt_model, pt_name, data, transpose):
+    """
+    Given a PyTorch model, the attribute's name and some data pointer (memoryview) Set the corresponding PT tensor with
+    the data (and eventual tranpose)
+    """
+    import torch
+
+    ptr = pt_model
+    for name in pt_name.split("."):
+        ptr = getattr(ptr, name)
+
+    array = np.array(data)
+    array = apply_transpose(array, transpose, ptr)
+    ptr.data = torch.from_numpy(array)
+
+
+def apply_transpose(array, transpose, ptr):
+    """
+    Given a numpy array, a transpose enum, and the target tensor shape Attempt to modify the array into something
+    fitting the target tensor The direction here is TF -> PT
+    """
+    if transpose is TransposeType.CONV2D:
+        # Conv2D weight:
+        #    TF: (kernel[0], kernel[1], num_in_channel, num_out_channel)
+        # -> PT: (num_out_channel, num_in_channel, kernel[0], kernel[1])
+        array = np.transpose(array, axes=(3, 2, 0, 1))
+    elif transpose is TransposeType.CONV1D:
+        # Conv1D weight:
+        #    TF: (kernel, num_in_channel, num_out_channel)
+        # -> PT: (num_out_channel, num_in_channel, kernel)
+        array = np.transpose(array, axes=(2, 1, 0))
+    elif transpose is TransposeType.SIMPLE:
+        array = np.transpose(array)
+
+    if len(ptr.shape) < len(array.shape):
+        array = array.squeeze()
+    elif len(ptr.shape) > len(ptr.shape):
+        array = np.expand_dims(array, axis=0)
+    if list(ptr.shape) != list(array.shape):
+        squeezed_tensor_shape = [dim for dim in ptr.shape if dim != 1]
+        squeezed_array_shape = [dim for dim in array.shape if dim != 1]
+        # Making sure we don't do wrong reshape
+        if squeezed_tensor_shape == squeezed_array_shape:
+            try:
+                array = array.reshape(ptr.shape)
+            except AssertionError as e:
+                raise e
```

Comment:
This is a personal addition to prevent bad reshapes from ever happening (Like `zeros((2, 3)).reshape((3, 2))` which would be a logic error.

The follow up PR, completely removes the need for this code (no random squeeze/unsqueeze either)

Reply:
> This is a personal addition to prevent bad reshapes from ever happening (Like zeros((2, 3)).reshape((3, 2)) which would be a logic error.

I don't think it ever happened in the tests at least but it felt like blindly reshaping could cause issues.

Reply:
Silently ignoring weights is absolutely not the right move, having spent several hours debugging things like that. And saying "it's going to be fixed in that other PR" for functionality at core is not an answer either.

Reply:
> Silently ignoring weights is absolutely not the right move

This is exactly what this prevents. It prevents silent errors (logic errors) and forces them to be real errors. 
The final `assert` of this function does the real erroring (Regardless of the attempt of reshaping) (I know we moved from `assert` to `if ... raise ValueError` I kept it to be closer to the original, but happy to make it `if raise ValueError`.

> "it's going to be fixed in that other PR" for functionality at core is not an answer either.

I don't think I claimed other PR solves this, I said it does not require the code anymore, meaning there's a single point of failure if tensors don't match. What I actually meant, is maybe we can follow the same path (so taking part of the other PR, and make it this one's code).

I can actually follow the same route, and make `apply_transpose` only about tensor reshaping and make validation another step as they are 2 distinct responsibility.

Reply:
At the time of my review, there was no final assert (see the [diff](https://github.com/huggingface/transformers/pull/16173/files/1ed3972bf3520f2b846f48fc8aac6342261c1dd0#diff-67993ece845388c9a9a1f342f4c82a2ed7f790f454134c39c3a63146616ea37a), so this comment is fixed now.

Reply:
My bad then ! Probably bad copy pasting from me early on.